zebra-state/src/service/finalized_state.rs

```diff
@@ -39,16 +43,36 @@ pub struct FinalizedState {
 impl FinalizedState {
     pub fn new(config: &Config, network: Network) -> Self {
         let (path, db_options) = config.db_config(network);
+
+        // Reject updates to existing values in column families
+        // with the following Merge rules:
+        // - Never: don't update existing values or delete keys, or
+        // - Delete: don't update existing values, but allow key deletion.
+        //
+        // For details see:
+        // https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#rocksdb-data-structures
+        let mut reject_updates = db_options.clone();
+        // the merge operator name is saved in the database
+        // changing the name requires a Zebra database version increment
+        reject_updates.set_merge_operator_associative(
+            "reject_updates_to_existing_values",
+            reject_updates_to_existing_values,
+        );
+
         let column_families = vec![
-            rocksdb::ColumnFamilyDescriptor::new("hash_by_height", db_options.clone()),
-            rocksdb::ColumnFamilyDescriptor::new("height_by_hash", db_options.clone()),
-            rocksdb::ColumnFamilyDescriptor::new("block_by_height", db_options.clone()),
-            rocksdb::ColumnFamilyDescriptor::new("tx_by_hash", db_options.clone()),
-            rocksdb::ColumnFamilyDescriptor::new("utxo_by_outpoint", db_options.clone()),
-            rocksdb::ColumnFamilyDescriptor::new("sprout_nullifiers", db_options.clone()),
-            rocksdb::ColumnFamilyDescriptor::new("sapling_nullifiers", db_options.clone()),
-            rocksdb::ColumnFamilyDescriptor::new("orchard_nullifiers", db_options.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("hash_by_height", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("height_by_hash", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("block_by_height", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("tx_by_hash", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("utxo_by_outpoint", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("sprout_nullifiers", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("sapling_nullifiers", reject_updates.clone()),
+            rocksdb::ColumnFamilyDescriptor::new("orchard_nullifiers", reject_updates),
+            // value_pools allows Update merges:
+            // chain value pool balances get updated for each block
+            //rocksdb::ColumnFamilyDescriptor::new("value_pools", db_options.clone()),
```

Comment:
Question: this will be uncommented in a separate PR, right? Would it make sense to have a `TODO` here, or is it going to be uncommented soon?

Reply:
I'll add a TODO when I do the revisions

Reply:
86f6ab34

Reply:
This change is obsolete, because the initial design didn't work.